### PR TITLE
feat: implement sponge hash implementation for poseidon

### PR DESCRIPTION
**Pull request by bajpai244**
This PR introduces an implementation of sponge hashing for our Poseidon implementation.

It also adds tests for the same, where the test vectors are borrowed from [go-iden3-crypto.](https://github.com/iden3/go-iden3-crypto/blob/edc36bfa52472e888b403337ba7941a122e99ab3/poseidon/poseidon_test.go#L137)
---------–––––––––––––––––––––––––––––––––––––––––––––––––––––––––

NOTE:  { this has been [resolved ](https://github.com/iden3/js-crypto/pull/2#issuecomment-1346908552) }

<del>In the test vectors, we directly use the bytes as input instead of using hex string and then converting them to bytes as in go-implementation.</del>

<del>This is done to save time for now as hex encoding, decoding was creating a few mismatches between go and js implementation, once that is resolved then we can follow the same scheme of using hex strings and converting them to bytes.</del>

<del>For now, this should not be a hindrance to using this implementation.</del>

### CODE REVIEW

File: src/poseidon/poseidon-opt.ts [hunk 1]
```diff
@@ -65,5 +67,50 @@ export class Poseidon {
...
+          inputs[j] = BigInt(0);
+        }
+        k = 1;
+      } else {
+        k += 1;
+      }
+    }
+
+    if (msg.length % SPONGE_CHUNK_SIZE != 0) {
+      let buff = new Uint8Array(new ArrayBuffer(SPONGE_CHUNK_SIZE));
```
Comment (Kolezhniuk) [inline src/poseidon/poseidon-opt.ts:98]: new Uint8Array(SPONGE_CHUNK_SIZE) ?

File: src/poseidon/poseidon-opt.ts [hunk 2]
```diff
@@ -65,5 +67,50 @@ export class Poseidon {
...
+
+    let k = 0;
+    for (let i = 0; i < parseInt(`${msg.length / SPONGE_CHUNK_SIZE}`); i += 1) {
+      dirty = true;
+      inputs[k] = utils.beBuff2int(msg.slice(SPONGE_CHUNK_SIZE * i, SPONGE_CHUNK_SIZE * (i + 1)));
+      if (k === SPONGE_INPUTS - 1) {
+        hash = this.hash(inputs);
+        dirty = false;
+        inputs = [];
+        inputs[0] = hash.valueOf();
```
Comment (Kolezhniuk) [inline src/poseidon/poseidon-opt.ts:87]: I guess the more elegant solution, what do you think?
```
inputs = new Array(SPONGE_INPUTS).fill(BigInt(0));
inputs[0] = inputs[0] = hash.valueOf();
```

File: src/poseidon/poseidon-opt.ts [hunk 3]
```diff
@@ -65,5 +67,50 @@ export class Poseidon {
 
     return F.normalize(state[0]);
   }
+
+  hashBytes(msg: Uint8Array): bigint {
+    let inputs: Array<bigint> = [];
+    for (let i = 0; i < SPONGE_INPUTS; i += 1) {
```
Comment (Kolezhniuk) [inline src/poseidon/poseidon-opt.ts:73]: inputs = new Array(SPONGE_INPUTS).fill(BigInt(0));

File: package.json [hunk 4]
```diff
@@ -45,7 +45,9 @@
     "typescript": "^4.9.3"
   },
   "dependencies": {
+    "@types/ramda": "^0.28.20",
     "blake-hash": "^2.0.0",
-    "ffjavascript": "^0.2.57"
+    "ffjavascript": "^0.2.57",
+    "ramda": "^0.28.0"
```
Comment (Kolezhniuk) [inline package.json:51]: Why do we need ramda?

**Review by Kolezhniuk — CHANGES_REQUESTED** [review]

File: package.json (hunk 4 above)
Comment (bajpai244) [reply package.json]: Ramda was used earlier in an util I was implementing, but now has been [removed in this commit.](https://github.com/iden3/js-crypto/pull/2/commits/26ea7d6fc3a89ee5c90cd257ec9c6f17c766c699)

File: src/poseidon/poseidon-opt.ts (hunk 1 above)
Comment (bajpai244) [reply src/poseidon/poseidon-opt.ts]: Amazing,  just checked the docs and wasn't aware of this constructor variant, always thought you need an ArrayBuffer or Array to construct them :)

Changed in [this](https://github.com/iden3/js-crypto/pull/2/commits/5cc7f47b45c5d8f19ab6092311b9bcb34e30642a) commit.

File: src/poseidon/poseidon-opt.ts (hunk 2 above)
Comment (bajpai244) [reply src/poseidon/poseidon-opt.ts:83]: Great! Just saw in my code that inputs are already initialized to and hence `input = [] ` is redundant.

Fixed in [this](https://github.com/iden3/js-crypto/pull/2/commits/5cc7f47b45c5d8f19ab6092311b9bcb34e30642a) commit.

**Comment by bajpai244** [comment]
> Could you fix the linter, please?

They have been fixed now

**Comment by bajpai244** [comment]
Update:

The test vectors have now been [updated](https://github.com/iden3/js-crypto/blob/5cc7f47b45c5d8f19ab6092311b9bcb34e30642a/tests/poseidon.test.ts#L28) and are directly using the hex strings as inputs instead of taking bytes.

File: package.json [hunk 5]
```diff
@@ -45,7 +45,10 @@
     "typescript": "^4.9.3"
   },
   "dependencies": {
+    "@types/bytebuffer": "^5.0.44",
+    "@types/ramda": "^0.28.20",
```
Comment (Kolezhniuk) [inline package.json:49]: remove  "@types/ramda": "^0.28.20",

File: src/poseidon/poseidon-opt.ts (hunk 2 above)
Comment (Kolezhniuk) [reply src/poseidon/poseidon-opt.ts:83]: please replace also
```
for (let j = 1; j < SPONGE_INPUTS; j += 1) {
          inputs[j] = BigInt(0);
 }

File: package.json [hunk 6]
```diff
@@ -45,7 +45,10 @@
...
     "blake-hash": "^2.0.0",
+    "bytebuffer": "^5.0.1",
```
Comment (Kolezhniuk) [inline package.json:51]: if you are using bytebuffer only for tests, please add this package and @types/bytebuffer into devDependencies of package.json, not direct dependencies.

**Review by Kolezhniuk — CHANGES_REQUESTED** [review]

File: package.json (hunk 4 above)
Comment (bajpai244) [reply package.json]: removed [here.](https://github.com/iden3/js-crypto/pull/2/commits/5b5bf5f7cb363f86ddde79ea2313b3d89bb8bf93)

File: package.json (hunk 6 above)
Comment (bajpai244) [reply package.json]: Done. [link to commit](https://github.com/iden3/js-crypto/pull/2/commits/5b5bf5f7cb363f86ddde79ea2313b3d89bb8bf93)

**Comment by bajpai244** [comment]
@Kolezhniuk  [This](https://github.com/iden3/js-crypto/pull/2#discussion_r1046136923) is resolved. [link here](https://github.com/iden3/js-crypto/pull/2/commits/bbf38ed86e2228a68f30fa279222534bf9546e79)

**Review by Kolezhniuk — APPROVED** [review]

**Pull request merged**